apps/browser/qml/pages/components/TabGridView.qml

```diff
@@ -0,0 +1,141 @@
+/****************************************************************************
+**
+** Copyright (c) 2014 - 2019 Jolla Ltd.
+** Copyright (c) 2019 - 2021 Open Mobile Platform LLC.
+**
+****************************************************************************/
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import QtQuick 2.1
+import Sailfish.Silica 1.0
+import org.nemomobile.configuration 1.0
+import "." as Browser
+
+SilicaGridView {
+    id: tabView
+
+    property bool portrait
+    property bool privateMode
+    property bool closingAllTabs
+
+    property var remorsePopup
+    readonly property bool largeScreen: Screen.sizeCategory > Screen.Medium
+    readonly property real thumbnailHeight: largeScreen
+                                            ? Screen.width / 3
+                                            : !portrait ? parent.height / 2.5 : parent.width / 1.66
+    readonly property real columns: largeScreen
+                                        ? portrait ? 2 : 3
+                                        : parent.width < 2 * parent.height
+                                          ? parent.width <= height ? 1 : 2 : 3
+    readonly property real thumbnailWidth: (parent.width - Theme.horizontalPageMargin * 2 - (Theme.paddingLarge * (columns - 1))) / columns
+
+    signal hide
+    signal enterNewTabUrl
+    signal activateTab(int index)
+    signal closeTab(int index)
+    signal closeAll
+    signal closeAllCanceled
+    signal closeAllPending
+
+    onCountChanged: if (count > 0) closingAllTabs = false
+    onClosingAllTabsChanged: if (closingAllTabs) closeAllPending()
+
+    width: parent.width - Theme.horizontalPageMargin
+    height: parent.height
+    x: Theme.horizontalPageMargin
```

Comment:
It turned out to be problematic to fix, besides, this problem will go away when switching to the popup menu

Reply:
The convention is to put the menu above the header. The tab header background rectangle is going to create a bit of a problem there because it would need to be below the pull down menu so it doesn't occlude the highlight.

But for reference the way to position the menu at the top of the window would be to assign the tabs grid view to the TabItem flickable property and add a header item to the grid view that is the same height as TabView header.

```
Private.TabItem {
    flickable: _persistentTabView
    TabGridView {
         id: _persistentTabView
         header: Item {
              width: 1
              height: tabs.headerHeight
         }
   }
}
```

The horizontal offset might still be a problem even then though. Provided there's no clipping though, you might be able to fix it by adding a negative x to the PullDownMenu as a counter margin and increasing the width of the menu to resolve the margin on the right.

Alternatively including the margin in the GridView delegates instead of the view itself would work too.

Reply:
Thanks for your suggestions for improvement. Fixed gap. Moved the pulley menu above the title, but this can only be considered as a temporary solution before the implementation of the popup menu